tests/lean/run/constProp.lean

```diff
@@ -237,14 +237,14 @@ notation:60 "(" σ ", " s ")"  " ⇓ " σ':60 => Bigstep σ s σ'
 theorem Bigstem.det (h₁ : (σ, s) ⇓ σ₁) (h₂ : (σ, s) ⇓ σ₂) : σ₁ = σ₂ := by
   induction h₁ generalizing σ₂ <;> cases h₂ <;> simp_all
   -- The rest of this proof should be automatic with congruence closure and a bit of forward reasoning
-  case seq ih₁ ih₂ h₁ h₂ =>
+  case seq ih₁ ih₂ _ h₁ h₂ =>
```

Comment:
What is going on here? Did you investigate why you had to add these extra `_`s?

Reply:
This is a consequence of the change to `simp_all` in https://github.com/leanprover/lean4/pull/2334/files#diff-003c74b01f7b82f8997b033dcd3d35690025c53797c5f806c6db9182952281e8.

The problem we had to address here was that `simp_all` was changing the goal unnecessarily by asserting everything and then try to clear. The solution Mario came up with only asserts the hypotheses we actually want to assert, and as a consequence of this change the order of hypotheses after `simp_all` can sometimes change a bit.

I agree this isn't ideal, but two things to note:
1) This problem didn't reappear when fixing up Mathlib to adapt to this change.
2) Arguably if you call tactics (here `simp_all`) in between `cases ...` and `case ... =>`, the onus is probably on the proof author to double check the order of hypotheses, as it's playing with fire to do this at all.

Reply:
@semorrison Notice that you are already tracking whether things changed or not. You don't need to rely on whether the `mvarId` changed or nor.

Reply:
@leodemoura Are you referring to the `modified` flag here? That flag is not sufficient to determine whether the mvarId changed, because it only tracks changes to the hypotheses, not the goal. The main purpose of it is to ensure that if the hypotheses are not simplified then there is no needless mvarId changing, so that the mvarId check is accurate.

Reply:
(Just noting that the extra `_`s are much reduced relative to @leodemoura's first query of this issue. Now there's just the one, instead of one in each of the cases here. There's still one, of course! Also noting that this issue does not seem to have arisen anywhere besides this file, including when we look at the [mathlib4 PR](https://github.com/leanprover-community/mathlib4/pull/6019) adapting to this change.)

Reply:
@semorrison It is great the issue is not affecting mathlib4, but mathlib is not the only user. Please modify the logic and ensure the order does not depend on whether a hypothesis was simplified or not. It feels like a very small change.

Reply:
@semorrison  this is strange, there should not be any `_` needed at all with the modification I suggested. "ensure the order does not depend on whether a hypothesis was simplified or not" should already be the case, and if not then there is a bug somewhere.

Reply:
@leodemoura, @digama0, I think there's some confusion here arising from the fact that it is the *old* `simp_all` which is reordering the hypotheses, while the *new* `simp_all` preserves the order exactly.

Thus the question becomes:
* do we prefer conserving the old behaviour (which was a consequence of re-asserting everything and then calling `tryClearMany` on everything) for backwards compatibility sake, or
* do we prefer the new behaviour, which strictly preserves the order of hypotheses when possible, including in this case.

Here are the relevant fragments of the goal:
* The `seq.seq` case immediately before `simp_all`:
```
case seq.seq
σ: State
s: Stmt
σ₁σ₁✝: State
s₁✝: Stmt
σ₂✝¹: State
s₂✝: Stmt
σ₃✝: State
a✝³: (σ₁✝, s₁✝) ⇓ σ₂✝¹
a✝²: (σ₂✝¹, s₂✝) ⇓ σ₃✝
a_ih✝¹: ∀ {σ₂ : State}, (σ₁✝, s₁✝) ⇓ σ₂ → σ₂✝¹ = σ₂
a_ih✝: ∀ {σ₂ : State}, (σ₂✝¹, s₂✝) ⇓ σ₂ → σ₃✝ = σ₂
σ₂σ₂✝: State
a✝¹: (σ₁✝, s₁✝) ⇓ σ₂✝
a✝: (σ₂✝, s₂✝) ⇓ σ₂
⊢ σ₃✝ = σ₂
```

Then after the new `simp_all` (note no changes):
```
case seq.seq
σ: State
s: Stmt
σ₁σ₁✝: State
s₁✝: Stmt
σ₂✝¹: State
s₂✝: Stmt
σ₃✝: State
a✝³: (σ₁✝, s₁✝) ⇓ σ₂✝¹
a✝²: (σ₂✝¹, s₂✝) ⇓ σ₃✝
a_ih✝¹: ∀ {σ₂ : State}, (σ₁✝, s₁✝) ⇓ σ₂ → σ₂✝¹ = σ₂
a_ih✝: ∀ {σ₂ : State}, (σ₂✝¹, s₂✝) ⇓ σ₂ → σ₃✝ = σ₂
σ₂σ₂✝: State
a✝¹: (σ₁✝, s₁✝) ⇓ σ₂✝
a✝: (σ₂✝, s₂✝) ⇓ σ₂
⊢ σ₃✝ = σ₂
```

Alternatively, the behaviour of `simp_all` on master (note that `σ₂σ₂✝: State` have migrated upwards, and hence out of the range where we need to specify them as arguments to `case seq`).
```
case seq.seq
σ: State
s: Stmt
σ₁σ₁✝: State
s₁✝: Stmt
σ₂✝¹: State
s₂✝: Stmt
σ₃✝σ₂σ₂✝: State
a✝³: (σ₁✝, s₁✝) ⇓ σ₂✝¹
a✝²: (σ₂✝¹, s₂✝) ⇓ σ₃✝
a_ih✝¹: ∀ {σ₂ : State}, (σ₁✝, s₁✝) ⇓ σ₂ → σ₂✝¹ = σ₂
a_ih✝: ∀ {σ₂ : State}, (σ₂✝¹, s₂✝) ⇓ σ₂ → σ₃✝ = σ₂
a✝¹: (σ₁✝, s₁✝) ⇓ σ₂✝
a✝: (σ₂✝, s₂✝) ⇓ σ₂
⊢ σ₃✝ = σ₂
```

Reply:
In particular, the request:

> Please modify the logic and ensure the order does not depend on whether a hypothesis was simplified or not.

is I think actually satisfied better by the new logic (which preserves order whether or not a hypothesis was simplified) than the old logic (which would sometimes change the order, whether or not a hypothesis was simplified).

Reply:
I will add some comments in the code explaining all this.

Reply:
I have added a simple test checking that `simp_all` now preserves order of hypothesis (which used to fail), and updated `RELEASE.md` to properly describe the new behaviour.